stellar-contract-env-common/src/val_wrapper.rs

```diff
@@ -0,0 +1,151 @@
+            }
+        }
+        impl AsMut<RawVal> for $tagname {
+            fn as_mut(&mut self) -> &mut RawVal {
+                &mut self.0
+            }
+        }
+        // Basic conversion support: wrapper to raw, and try-into helper.
+        impl From<$tagname> for RawVal {
+            fn from(b: $tagname) -> Self {
+                b.0
+            }
+        }
+        impl crate::RawValConvertible for $tagname {
+            #[inline(always)]
+            fn is_val_type(v: RawVal) -> bool {
+                v.has_tag(Tag::$tagname)
+            }
+            #[inline(always)]
+            unsafe fn unchecked_from_val(v: RawVal) -> Self {
+                $tagname(v)
+            }
+        }
+
+        // EnvVal ref/mut/from support
+        impl<E: Env> AsRef<$tagname> for EnvVal<E, $tagname> {
+            fn as_ref(&self) -> &$tagname {
+                &self.val
+            }
+        }
+        impl<E: Env> AsMut<$tagname> for EnvVal<E, $tagname> {
+            fn as_mut(&mut self) -> &mut $tagname {
+                &mut self.val
+            }
+        }
+        impl<E: Env> From<EnvVal<E, $tagname>> for EnvVal<E, RawVal> {
+            fn from(ev: EnvVal<E, $tagname>) -> Self {
+                EnvVal {
+                    env: ev.env,
+                    val: ev.val.to_raw(),
+                }
+            }
+        }
+        impl<E: Env> From<EnvVal<E, $tagname>> for RawVal {
+            fn from(ev: EnvVal<E, $tagname>) -> Self {
+                ev.val.0
+            }
+        }
+        impl<E: Env> From<EnvVal<E, $tagname>> for $tagname {
+            fn from(ev: EnvVal<E, $tagname>) -> Self {
+                ev.val
+            }
+        }
+        impl<E: Env> EnvVal<E, $tagname> {
+            pub fn as_raw(&self) -> &RawVal {
+                self.val.as_ref()
+            }
+            pub fn to_raw(&self) -> RawVal {
+                self.val.to_raw()
+            }
+        }
+
+        // wasmi / VM argument support
+        #[cfg(feature = "vm")]
+        impl wasmi::FromValue for $tagname {
+            fn from_value(val: wasmi::RuntimeValue) -> Option<Self> {
+                let maybe: Option<u64> = val.try_into();
+                match maybe {
+                    Some(u) => {
+                        let raw = RawVal::from_payload(u);
+                        if <Self as crate::RawValConvertible>::is_val_type(raw) {
+                            Some(unsafe {
+                                <Self as crate::RawValConvertible>::unchecked_from_val(raw)
+                            })
+                        } else {
+                            None
+                        }
+                    }
+                    None => None,
+                }
+            }
+        }
+        #[cfg(feature = "vm")]
+        impl From<$tagname> for wasmi::RuntimeValue {
+            fn from(v: $tagname) -> Self {
+                wasmi::RuntimeValue::I64(v.as_raw().get_payload() as i64)
```

Comment:
What's the use case for these conversions to/from `wasmi::RuntimeValue`? So that we can specify them directly as the host function argument types instead of `RawVal`? 
Also, why is the `from_value` expecting an `u64` while the `from` converts the payload to an `i64`?